frontend/src/components/auth/login-dialog.tsx

```diff
@@ -0,0 +1,168 @@
+import {
+  RegisterEmailAlreadyExistsError,
+  RegisterUsernameAlreadyTakenError,
+  useRegister,
+} from '@/hooks/auth/useRegister';
+import {
+  useResendVerificationCode,
+  useVerifySignup,
+  VerificationCodeExpiredError,
+  VerificationCodeInvalidError,
+} from '@/hooks/auth/useVerify';
+import { LoginUser, RegisterUser, VerifyUserCode } from '@/types/auth';
+import { useState } from 'react';
+import { toast } from 'sonner';
+
+export default function LoginDialog() {
+  const [isOpen, setIsOpen] = useState(false);
+  const [activeTab, setActiveTab] = useState('signup');
+  const loginMutation = useLogin();
+  const registerMutation = useRegister();
+
+  const [isUserAwaitingEmailVerification, setIsUserAwaitingEmailVerification] =
+    useState<LoginUser | null>(null);
+
+  const verifySignupMutation = useVerifySignup();
+  const resendVerificationCodeMutation = useResendVerificationCode();
+  const [initialCodeSentAt, setInitialCodeSentAt] = useState<number | null>(
+    null
+  );
+
+  const handleLogin = async (data: LoginUser) => {
+    try {
+      await loginMutation.mutateAsync(data);
+      setIsOpen(false);
+      toast.success('Login successful');
+      // Refresh the page after successful login
+      window.location.reload();
+    } catch (error) {
+      if (error instanceof UnverifiedAccountError) {
+        setIsUserAwaitingEmailVerification({
+          email: data.email,
+          password: data.password,
+        });
+        setInitialCodeSentAt(Date.now());
+      } else {
+        console.error('Login failed:', error);
+        toast.error('Login failed');
+      }
+    }
+  };
+
+  const handleSignup = async (data: RegisterUser) => {
+    try {
+      await registerMutation.mutateAsync(data);
+      toast.success('Signup successful');
+      setIsUserAwaitingEmailVerification({
+        email: data.email,
+        password: data.password,
+      });
+      setInitialCodeSentAt(Date.now());
+    } catch (error) {
+      if (error instanceof RegisterUsernameAlreadyTakenError) {
+        toast.error('Username already taken');
+      } else if (error instanceof RegisterEmailAlreadyExistsError) {
+        toast.error('Email already exists');
+      } else {
+        console.error('Signup failed:', error);
+        toast.error('Signup failed');
+      }
+    }
+  };
+
+  const handleVerifySignup = async (data: VerifyUserCode) => {
+    if (!isUserAwaitingEmailVerification) return;
+
+    try {
+      await verifySignupMutation.mutateAsync({
+        email: isUserAwaitingEmailVerification.email,
+        verificationCode: data.verificationCode,
+      });
+      toast.success('Email verified.');
+      handleLogin({
+        email: isUserAwaitingEmailVerification.email,
+        password: isUserAwaitingEmailVerification.password,
+      });
+    } catch (error) {
+      if (error instanceof VerificationCodeInvalidError) {
+        toast.error('Invalid verification code');
```

Comment:
Since invalid password input highlights the input box and has the error message in red underneath it. 
In addition to showing the toast error, the corresponding input field (username or email) could maybe be visually highlighted (e.g., by changing the border color to red) to draw attention to the required change. 
Just a minor suggestion